File: packages/jupytercad-extension/src/worker/geometry/geomCircle.ts
import { IGeomCircle } from '../../_interface/geomCircle';
import { getOcc } from '../actions';

export function _GeomCircle(arg: IGeomCircle): any {
  const oc = getOcc();
  const center = new oc.gp_Pnt_3(arg.CenterX, arg.CenterY, arg.CenterZ);
  const norm = new oc.gp_Dir_4(arg.NormalX, arg.NormalY, arg.NormalZ);
  const radius = arg.Radius;
  const circle = new oc.GC_MakeCircle_6(center, norm, radius).Value();
  const edge = new oc.BRepBuilderAPI_MakeEdge_8(circle.get().Circ()).Edge();
  const circleWire = new oc.BRepBuilderAPI_MakeWire_2(edge).Wire();

  return circleWire;
}

File: jupytercad/freecad/props/geometry/geom_circle.py
from typing import Any, Dict

from ...tools import redirect_stdout_stderr

from ..base_prop import BaseProp

with redirect_stdout_stderr():
    try:
        import freecad as fc
        import Part
    except ImportError:
        fc = None


class Part_GeomCircle(BaseProp):
    @staticmethod
    def name() -> str:
        return "Part::GeomCircle"

    @staticmethod
    def fc_to_jcad(prop_value: Any, **kwargs) -> Any:
        center = prop_value.Center
        radius = prop_value.Radius
        angle = prop_value.AngleXU
        normal = prop_value.Axis
        return {
            "TypeId": Part_GeomCircle.name(),
            "CenterX": center.x,
            "CenterY": center.y,
            "CenterZ": center.z,
            "NormalX": normal.x,
            "NormalY": normal.y,
            "NormalZ": normal.z,
            "AngleXU": angle,
            "Radius": radius,
        }

    @staticmethod
    def jcad_to_fc(prop_value: Dict, fc_object: Any, **kwargs) -> Any:
        if not fc:
            return
        Center = fc.app.Base.Vector(
            prop_value["CenterX"],
            prop_value["CenterY"],
            prop_value["CenterZ"],
        )

        Axis = fc.app.Base.Vector(
            prop_value["NormalX"],
            prop_value["NormalY"],
            prop_value["NormalZ"],
        )

        Radius = prop_value["Radius"]
        if fc_object:
            fc_object.Center = Center
            fc_object.Axis = Axis
            fc_object.AngleXU = prop_value["AngleXU"]
            fc_object.Radius = Radius
            return None
        else:
            return Part.Circle(Center, Axis, Radius)
